arcgis-java-toolkit/src/main/java/com/esri/arcgisruntime/toolkit/UtilityNetworkTraceTool.java

```diff
@@ -0,0 +1,999 @@
+   * selected utility network and the starting point does not already exist.
+   *
+   * @param feature the feature to use as the basis for the starting point
+   * @throws NullPointerException if feature is null
+   * @since 100.15.0
+   */
+  public void addStartingPoint(ArcGISFeature feature) {
+    addStartingPoint(feature, null);
+  }
+
+  /**
+   * Adds a starting point to be used for a trace from the provided feature and point.
+   *
+   * <p>
+   * This enables the adding of starting points programmatically, and not just via clicks on the MapView.
+   * For example, feature results from a query, selection, geocode, route, geoprocessing, search, and more.
+   *
+   * <p>
+   * A starting point is only added to the list if a Utility Network has been selected, the feature is part of the
+   * selected utility network and the starting point does not already exist.
+   *
+   * @param feature the feature to use as the basis for the starting point
+   * @param mapPoint the point to use - if it's polyline for location along the line
+   * @throws NullPointerException if feature is null
+   * @since 100.15.0
+   */
+  public void addStartingPoint(ArcGISFeature feature, Point mapPoint) {
+    if (selectedUtilityNetworkProperty != null) {
+      var geometry = feature.getGeometry();
+      UtilityElement utilityElement = null;
+
+      try {
+        utilityElement = selectedUtilityNetworkProperty.get().createElement(feature);
+      } catch (Exception e) {
+        // if a feature does not belong to the selected utility network it is ignored
+      }
+
+      if (utilityElement != null) {
+        // check if the starting point already exists
+        var startingPointAlreadyExists = false;
+        var utilityElementId = utilityElement.getObjectId();
+        for (UtilityNetworkTraceStartingPoint startingPoint : startingPointsProperty) {
+          if (startingPoint.getUtilityElement().getObjectId() == utilityElementId) {
+            startingPointAlreadyExists = true;
+          }
+        }
+
+        // only continue if the starting point does not already exist
+        if (!startingPointAlreadyExists) {
+          if (utilityElement.getNetworkSource().getSourceType() ==
+            UtilityNetworkSource.Type.EDGE && geometry instanceof Polyline) {
+            // configure edge utility elements
+            Polyline polyline = (Polyline) geometry;
+            if (polyline.hasZ()) {
+              // get the geometry of the identified feature as a polyline, and remove the z component
+              polyline = (Polyline) GeometryEngine.removeZ(polyline);
+            }
+            if (mapPoint != null && mapPoint.getSpatialReference() != polyline.getSpatialReference()) {
+              polyline = (Polyline) GeometryEngine.project(polyline, mapPoint.getSpatialReference());
+            }
+            geometry = polyline;
+
+            // compute how far the clicked location is along the edge feature
```

Comment:
Comment says "clicked" but this is a public method so map point is not necessarily a click point.